File: sanghyun/programmers-2/땅따먹기.java
class Solution {
    int solution(int[][] land) {
        int answer = 0;  // 최종 결과로 반환할 값
        
        for (int i = 1; i < land.length; i++) {  // 행 탐색 (첫 번째 행은 무시)
            for (int j = 0; j < land[i].length; j++) {  // 열 탐색
                int max = 0;  // 이전 행에서 현재 열을 제외한 최대값 저장 변수
                for (int k = 0; k < land[i-1].length; k++) {  // 이전 행의 열들을 순회
                    if (k != j) {  // 현재 열을 제외한 경우에만
                        max = Math.max(max, land[i-1][k]);  // 최대값 갱신
                    }
                }
                land[i][j] = land[i][j] + max;  // 현재 위치에 최대값을 더하여 저장
            }
        }
        
        for (int i = 0; i < land[0].length; i++) {  // 마지막 행의 열을 순회
            answer = Math.max(answer, land[land.length-1][i]); 
        }
        return answer; 
    }
}

File: soyeong/programmers-2/땅따먹기.py
def solution(land):
    answer = 0

    # 한 행에 한칸만, 그리고 다음 행을 넘어갈때 같은 열을 밟을 수 없다는 사항이 중요
    # (1,0)위치를 예로 들면 (0,1)(0,2)(0,3) 위치에서만 (1,0)은 밟을 수 있고 (0,0)은 (1,0)을 밟을 수 없음
    # 각 위치의 다양한 경우의 수 따져 가장 큰 수를 찾는 케이스니 DP 사용

    # land의 DP 값을 저장하도록 2차원 배열 정의
    dp = [[0 for j in range(len(land[0]))] for i in range(len(land))]

    # 0행은 먼저 저장 for문은 1행부터 시작
    dp[0] = land[0]

    for i in range(1, len(dp)):  # 행
        for j in range(len(dp[0])):  # 열
            # 같은 열인 것을 제외하는 방법 (j열만 제외하고 배열 붙이기)
            dp[i][j] = land[i][j] + max(dp[i - 1][:j] + dp[i - 1][j + 1:])

    return max(dp[-1])  # 마지막 열에서 가장 큰 값 출력